### PR TITLE
feat(dot/peerset): Implement peer scoring

### LINKED ISSUE #1399: research and implement scoring for network sync module
## Task summary

- scoring should be used when determining who to try to sync from first
- basic premise is score goes up when you successfully receive a response from a peer, score goes down otherwise
- research how to ideally implement this so that peers we are more likely to get a response from are prioritized without completely ignoring other/new peers

### PR DESCRIPTION
**Pull request by arijitAD**
## Changes

- [x] Implement Peerset
- [x] Implement peer state manager
- [x] Implement Network behaviour
- [x] Integrate with gossamer
- [x] Add test case

## Tests

```
go test github.com/ChainSafe/gossamer/dot/peerset -v 
```

## Issues

- closes #1399

## Primary Reviewer 

- @noot

### CODE REVIEW

File: dot/peerset/peerset.go [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+
+}
+
+// TODO: Analysis and implementation yet to be done for the usage of this in gossamer...
+func (ps *Peerset) onSetReservedPeer(setId int, peerId peer.ID) {
+	return
+}
+
+// Returns the list of reserved peers.
+func (ps *Peerset) reservedPeers(setId int) []peer.ID {
```
Comment (noot) [inline dot/peerset/peerset.go:355]: I think this is the same functionality as the current network "persistent peers"

Comment (arijitAD) [reply dot/peerset/peerset.go]: I have moved the persistent peers functionality to peerset.

File: dot/network/discovery.go [hunk 2]
```diff
@@ -208,16 +210,8 @@ func (d *discovery) findPeers(ctx context.Context) {
 
 			logger.Trace("found new peer via DHT", "peer", peer.ID)
 
-			// found a peer, try to connect if we need more peers
-			if len(d.h.Network().Peers()) < d.maxPeers {
```
Comment (noot) [inline dot/network/discovery.go:212]: why is this removed?

File: dot/peerset/peerset.go [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+	// Reject incoming connect request
+	Reject
+)
+
+// Message that can be sent by the peer set manager (PSM).
+type Message struct {
+	messageStatus MessageStatus
+	setID         uint64
+	peerID        peer.ID
+	IncomingIndex uint64 // TODO: make IncomingIndex a type of uint64.
```
Comment (noot) [inline dot/peerset/peerset.go:72]: is this still a TODO?

File: dot/peerset/peerset.go [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+		reservedOnly:  reservedOnly,
+	}
+
+	return &ConfigSet{[]*config{set}}
+}
+
+func newPeerSet(cfg *ConfigSet, actionCh <-chan Action) (*PeerSet, error) {
+	now := time.Now()
+
+	peerSate, err := NewPeerState(cfg.Set)
```
Comment (noot) [inline dot/peerset/peerset.go:160]: peerState*

File: dot/network/discovery.go (hunk 2 above)
Comment (noot) [reply dot/network/discovery.go:212]: it seems `AddToPeerSet` doesn't actually connect to the peer

Comment (arijitAD) [reply dot/network/discovery.go:212]: `maxPeers ` and `minPeers` are maintained in the peerset.

Comment (noot) [reply dot/network/discovery.go:212]: I can't find where in the code it actually connects to the peer, it seems all `connect` calls are removed in favour of `AddToPeerSet` which doesn't seem to connect (also integration tests are failing) - I think you probably need to leave the `connect` calls in the network package, as the peerset shouldn't deal with the actual networking, just tracking peer info

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

**Review by EclesioMeloJunior — DISMISSED** [review]

File: dot/core/messages.go [hunk 5]
```diff
@@ -59,6 +61,10 @@ func (s *Service) HandleTransactionMessage(msg *network.TransactionMessage) (boo
 			externalExt := types.Extrinsic(append([]byte{byte(types.TxnExternal)}, tx...))
 			val, err := rt.ValidateTransaction(externalExt)
 			if err != nil {
+				s.net.ReportPeer(peerset.ReputationChange{
+					Value:  peerset.BadTransactionValue,
+					Reason: peerset.BadTransactionReason,
+				}, peerID)
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/core/messages.go]: No, I have added this check
```
if errors.Is(err, runtime.ErrInvalidTransaction)
```

File: dot/network/connmgr.go [hunk 6]
```diff
@@ -90,18 +94,18 @@ func (*ConnManager) UpsertTag(peer.ID, string, func(int) int) {}
 func (*ConnManager) GetTagInfo(peer.ID) *connmgr.TagInfo { return &connmgr.TagInfo{} }
 
 // TrimOpenConns peer
-func (*ConnManager) TrimOpenConns(ctx context.Context) {}
+func (*ConnManager) TrimOpenConns(_ context.Context) {}
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/connmgr.go:97]: I have added this check for error checking
```
if errors.Is(err, runtime.ErrInvalidTransaction)
```

[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/connmgr.go:97]: I think we need to keep it otherwise interface will not be implemented.

File: dot/network/service.go [hunk 7]
```diff
@@ -743,3 +746,65 @@ func (*Service) StartingBlock() int64 {
...
+		logger.Debug("connection dropped successfully ", "peer", peerID)
+	}
+}
+
+func (s *Service) startProcessingMsg() {
+	msgCh := s.host.cm.peerSetHandler.Messages()
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/service.go:800]: This looks interesting. I will explore it.

File: dot/peerset/peerstate.go [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+	if err != nil {
+		return
+	}
+
+	switch n.state[idx] {
+	case ingoing:
+		ps.sets[idx].numIn++
+	case outgoing:
+		ps.sets[idx].numOut++
+	}
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate.go:276]: There are multiple types of states. Currently, we are using two. With the switch, the code feels cleaner.

File: dot/peerset/peerstate.go [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+// non slot occupying nodes don't count towards the number of slots.
+func (ps *PeersState) tryAcceptIncoming(setID int, peerID peer.ID) error {
+	var isNoSlotOccupied bool
+	if _, ok := ps.sets[setID].noSlotNodes[peerID]; ok {
+		isNoSlotOccupied = true
+	}
+
+	// if slot is not available and the node is not a reserved node then error
+	if ps.hasFreeIncomingSlot(setID) && !isNoSlotOccupied {
+		return errors.New("not enough incoming slots")
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate.go:393]: Done

File: dot/peerset/peerstate.go [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+		return err
+	}
+
+	if _, ok := info.noSlotNodes[peerID]; !ok {
+		switch n.state[idx] {
+		case ingoing:
+			info.numIn--
+		case outgoing:
+			info.numOut--
+		case notMember, notConnected:
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate.go:293]: Changes it to default.

File: dot/peerset/peerstate.go [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+	if !isNoSlotOccupied {
+		// this need to be added as incoming connection allocate slot.
+		ps.sets[setID].numIn++
+	}
+
+	return nil
+}
+
+// isConnected returns true if peer is connected else false
+func isConnected(state MembershipState) bool {
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate.go:412]: Done

File: dot/peerset/peerstate_test.go [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+import (
+	"testing"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFullSlotIn(t *testing.T) {
+	state := newTestPeerState(t, 1, 1)
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate_test.go:11]: Done

File: dot/peerset/test_helpers.go [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+	reservedPeer  = peer.ID("testReservedPeer")
+	reservedPeer2 = peer.ID("testReservedPeer2")
+	discovered1   = peer.ID("testDiscovered1")
+	discovered2   = peer.ID("testDiscovered2")
+	incomingPeer  = peer.ID("testIncoming")
+	incoming2     = peer.ID("testIncoming2")
+	incoming3     = peer.ID("testIncoming3")
+	peer1         = peer.ID("testPeer1")
+	peer2         = peer.ID("testPeer2")
+)
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/test_helpers.go:22]: Done

File: dot/network/state.go [hunk 14]
```diff
@@ -58,6 +59,38 @@ type Syncer interface {
...
+
+// PeerRemover is the interface used by the PeerSetHandler to remove peers from peerSet.
+type PeerRemover interface {
+	DisconnectPeer(int, ...peer.ID)
+	RemoveReservedPeer(int, ...peer.ID)
+	RemovePeer(int, ...peer.ID)
+}
+
+// PeerGetter is the interface used by the PeerSetHandler to get the peer data from peerSet.
+type PeerGetter interface {
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/state.go:92]: Done

File: dot/network/state.go [hunk 15]
```diff
@@ -58,6 +59,38 @@ type Syncer interface {
...
+// PeerAdder is the interface used by the PeerSetHandler to add peers in peerSet.
+type PeerAdder interface {
+	Incoming(int, ...peer.ID)
+	AddReservedPeer(int, ...peer.ID)
+	AddPeer(int, ...peer.ID)
+	SetReservedPeer(int, ...peer.ID)
+}
+
+// PeerRemover is the interface used by the PeerSetHandler to remove peers from peerSet.
+type PeerRemover interface {
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/state.go:85]: Done

File: dot/network/state.go [hunk 16]
```diff
@@ -58,6 +59,38 @@ type Syncer interface {
...
+	Start()
+	Stop()
+	ReportPeer(peerset.ReputationChange, ...peer.ID)
+	PeerAdder
+	PeerRemover
+	PeerGetter
+}
+
+// PeerAdder is the interface used by the PeerSetHandler to add peers in peerSet.
+type PeerAdder interface {
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/state.go:77]: Done

File: dot/peerset/peerstate.go [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+	maxIn uint32
+
+	// maximum allowed number of slot occupying nodes for which the MembershipState is outgoing.
+	maxOut uint32
+
+	// list of node identities (discovered or not) that don't occupy slots.
+	// Note for future readers: this module is purely dedicated to managing slots.
+	// If you are considering adding more features, please consider doing so outside this module rather
+	// than inside.
+	noSlotNodes map[peer.ID]interface{}
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate.go:53]: Done

File: dot/peerset/peerstate.go [hunk 18]
```diff
@@ -0,0 +1,414 @@
...
+		info := Info{
+			numIn:       0,
+			numOut:      0,
+			maxIn:       cfg.inPeers,
+			maxOut:      cfg.outPeers,
+			noSlotNodes: make(map[peer.ID]interface{}),
+		}
+
+		infoSet = append(infoSet, info)
+	}
```
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/peerset/peerstate.go:134]: Done

File: dot/network/discovery.go [hunk 19]
```diff
@@ -208,16 +210,12 @@ func (d *discovery) findPeers(ctx context.Context) {
...
-				return
-			}
+			d.h.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
+			d.handler.AddPeer(0, peer.ID)
 		}
 	}
 }
+
+func (d *discovery) findPeer(peerID peer.ID) (peer.AddrInfo, error) {
+	return d.dht.FindPeer(d.ctx, peerID)
```
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/network/discovery.go:220]: Well it depends, sometimes both are needed.

For example since you can't fiddle with Go's standard library HTTP goroutine dispatcher (and pass a context to each goroutine starting), you might want to cancel the context for all requests being processed when the entire server should shutdown, and so you need to set a context as field of your server/handler receiver struct.

**Review by edwardmack — APPROVED** [review]
